File: backend/Services/IoTestService.cs
using IO_Checkout_Tool.Constants;
using IO_Checkout_Tool.Models.Common;
using Shared.Library.Repositories.Interfaces;
using IO_Checkout_Tool.Services.Interfaces;
using IO_Checkout_Tool.SharedComponents;
using Shared.Library.Models.Entities;
using Shared.Library.DTOs;

namespace IO_Checkout_Tool.Services;

public class IoTestService : IIoTestService
{
    private readonly IIoRepository _ioRepository;
    private readonly ITestHistoryService _historyService;
    private readonly ICloudSyncService _cloudSyncService;
    private readonly ILogger<IoTestService> _logger;

    public IoTestService(
        IIoRepository ioRepository, 
        ITestHistoryService historyService,
        ICloudSyncService cloudSyncService,
        ILogger<IoTestService> logger)
    {
        _ioRepository = ioRepository;
        _historyService = historyService;
        _cloudSyncService = cloudSyncService;
        _logger = logger;
    }

    public async Task<Io?> GetNextUntestedTagAsync()
    {
        return await _ioRepository.GetNextUntestedAsync();
    }

    public async Task<bool> MarkTestPassedAsync(Io tag, string comments = "")
    {
        return await UpdateTestResultAsync(tag, TestConstants.RESULT_PASSED, comments);
    }

    public async Task<bool> MarkTestFailedAsync(Io tag, string comments)
    {
        return await UpdateTestResultAsync(tag, TestConstants.RESULT_FAILED, comments);
    }

    public async Task<bool> ClearTestResultAsync(Io tag)
    {
        try
        {
            var dbTag = await GetDbTagAsync(tag.Id);
            if (dbTag == null) return false;

            // If the tag is already fully cleared (no result and no comments), do nothing
            if (string.IsNullOrEmpty(dbTag.Result) && string.IsNullOrEmpty(dbTag.Comments))
            {
                _logger.LogDebug("Tag {TagId} is already cleared, skipping clear operation", tag.Id);
                return true;
            }

            // Store the old values to determine if we should add history
            var hadComments = !string.IsNullOrEmpty(dbTag.Comments);
            var hadResult = !string.IsNullOrEmpty(dbTag.Result);

            // Prepare the cleared message
            string? clearedComment = null;
            
            // Add local history if there were comments or a previous result
            if (hadComments || hadResult)
            {
                if (hadResult && hadComments)
                {
                    clearedComment = $"Previous result was {dbTag.Result}";
                }
                else if (hadResult)
                {
                    clearedComment = $"Previous result was {dbTag.Result}";
                }
                else
                {
                    clearedComment = "Cleared comments";
                }
                
                await _historyService.AddTestHistoryAsync(
                    dbTag.Id,
                    TestConstants.RESULT_CLEARED,
                    clearedComment,
                    tag.State ?? "");
            }
            
            ClearTagResult(dbTag);
            
            await _ioRepository.UpdateAsync(dbTag);
            
            // Always sync with "Cleared" result if there was something to clear
            if (hadComments || hadResult)
            {
                await SyncClearedToCloudAsync(dbTag, tag.State, TestConstants.RESULT_CLEARED, clearedComment);
            }
            else
            {
                // Sync without adding history - nothing was cleared
                await SyncToCloudAsync(dbTag, tag.State);
            }
            
            return true;
        }
        catch
        {
            return false;
        }
    }

    public async Task<CommentUpdateResult> UpdateCommentAsync(Io tag, string newComment)
    {
        try
        {
            var dbTag = await GetDbTagAsync(tag.Id);
            if (dbTag == null) return CommentUpdateResult.Failed("Tag not found in database");

            var oldComment = dbTag.Comments;
            
            // Normalize empty values for comparison
            var normalizedOldComment = string.IsNullOrEmpty(oldComment) ? "" : oldComment;
            var normalizedNewComment = string.IsNullOrEmpty(newComment) ? "" : newComment;
            
            // If there's no actual change, don't create history
            if (normalizedOldComment == normalizedNewComment)
            {
                return CommentUpdateResult.Successful(changesWereMade: false); // No change needed
            }
            
            dbTag.Comments = newComment;
            dbTag.Timestamp = CreateTimestamp();

            var (historyResult, historyComment) = DetermineCommentChange(oldComment, newComment);

            await _ioRepository.UpdateAsync(dbTag);
            await _historyService.AddTestHistoryAsync(dbTag.Id, historyResult, historyComment, tag.State);

            // Sync to cloud with the comment change result
            await SyncCommentChangeToCloudAsync(dbTag, tag.State, historyResult, historyComment);

            return CommentUpdateResult.Successful(changesWereMade: true);
        }
        catch (Exception ex)
        {
            return CommentUpdateResult.Failed($"Error updating comment: {ex.Message}");
        }
    }

    private async Task<bool> UpdateTestResultAsync(Io tag, string result, string comments)
    {
        try
        {
            var dbTag = await GetDbTagAsync(tag.Id);
            if (dbTag == null) return false;

            UpdateTagWithResult(dbTag, result, comments);
            await _ioRepository.UpdateAsync(dbTag);
            await _historyService.AddTestHistoryAsync(dbTag.Id, result, comments, tag.State);

            // Sync to cloud
            await SyncToCloudAsync(dbTag, tag.State);

            return true;
        }
        catch
        {
            return false;
        }
    }

    private async Task SyncToCloudAsync(Io io, string? state = null)
    {
        try
        {
            if (io.Id <= 0) return;

            // Get TestedBy from the most recent TestHistory record (not Environment.UserName)
            var history = await _historyService.GetHistoryForIoAsync(io.Id);
            var testedBy = history.FirstOrDefault()?.TestedBy ?? "Unknown";

            var update = new IoUpdateDto
            {
                Id = io.Id,
                Result = io.Result,
                Timestamp = io.Timestamp,
                Comments = io.Comments,
                TestedBy = testedBy, // Use actual user from TestHistory
                State = state,
                Version = io.Version
            };

            // Don't await - let it run in background
            _ = _cloudSyncService.SyncIoUpdateAsync(update).ContinueWith(task =>
            {
                if (!task.Result)
                {
                    _logger.LogWarning("Failed to sync IO {IoId} to cloud (will retry from queue)", io.Id);
                }
            }, TaskContinuationOptions.OnlyOnRanToCompletion);
        }
        catch (Exception ex)
        {
            // Log but don't fail the local operation
            _logger.LogError(ex, "Error syncing IO {IoId} to cloud", io.Id);
        }
    }

    private async Task SyncClearedToCloudAsync(Io io, string? state, string? historyResult, string? clearedComment)
    {
        try
        {
            if (io.Id <= 0) return;

            // Get TestedBy from the most recent TestHistory record (not Environment.UserName)
            var history = await _historyService.GetHistoryForIoAsync(io.Id);
            var testedBy = history.FirstOrDefault()?.TestedBy ?? "Unknown";

            var update = new IoUpdateDto
            {
                Id = io.Id,
                Result = historyResult,
                Timestamp = io.Timestamp,
                Comments = clearedComment,
                TestedBy = testedBy, // Use actual user from TestHistory
                State = state,
                Version = io.Version
            };

            // Don't await - let it run in background
            _ = _cloudSyncService.SyncIoUpdateAsync(update).ContinueWith(task =>
            {
                if (!task.Result)
                {
                    _logger.LogWarning("Failed to sync IO {IoId} to cloud (will retry from queue)", io.Id);
                }
            }, TaskContinuationOptions.OnlyOnRanToCompletion);
        }
        catch (Exception ex)
        {
            // Log but don't fail the local operation
            _logger.LogError(ex, "Error syncing IO {IoId} to cloud", io.Id);
        }
    }

    private async Task SyncCommentChangeToCloudAsync(Io io, string? state, string historyResult, string historyComment)
    {
        try
        {
            if (io.Id <= 0) return;

            var update = new IoUpdateDto
            {
                Id = io.Id,
                Result = historyResult,  // Send the comment change result type
                Timestamp = io.Timestamp,
                Comments = io.Comments,  // Send the actual current comment value, not the history text
                TestedBy = Environment.UserName,
                State = state,
                Version = io.Version
            };

            // Don't await - let it run in background
            _ = _cloudSyncService.SyncIoUpdateAsync(update).ContinueWith(task =>
            {
                if (!task.Result)
                {
                    _logger.LogWarning("Failed to sync IO {IoId} to cloud (will retry from queue)", io.Id);
                }
            }, TaskContinuationOptions.OnlyOnRanToCompletion);
        }
        catch (Exception ex)
        {
            // Log but don't fail the local operation
            _logger.LogError(ex, "Error syncing IO {IoId} to cloud", io.Id);
        }
    }

    private async Task<Io?> GetDbTagAsync(int tagId)
    {
        return await _ioRepository.GetByIdAsync(tagId);
    }

    private void UpdateTagWithResult(Io dbTag, string result, string comments)
    {
        var timestamp = CreateTimestamp();
        dbTag.Result = result;
        dbTag.Timestamp = timestamp;
        dbTag.Comments = comments;
    }

    private void ClearTagResult(Io dbTag)
    {
        var timestamp = CreateTimestamp();
        dbTag.Result = null;
        dbTag.Comments = "";
        dbTag.Timestamp = timestamp;
    }

    private string CreateTimestamp()
    {
        return DateTime.UtcNow.ToString(TestConstants.TIMESTAMP_FORMAT);
    }

    private (string historyResult, string historyComment) DetermineCommentChange(string? oldComment, string? newComment)
    {
        var historyResult = DetermineCommentChangeType(oldComment, newComment);
        var historyComment = CreateCommentHistoryText(oldComment, newComment);
        return (historyResult, historyComment);
    }

    private string DetermineCommentChangeType(string? oldComment, string? newComment)
    {
        if (string.IsNullOrEmpty(oldComment) && !string.IsNullOrEmpty(newComment))
            return TestConstants.RESULT_COMMENT_ADDED;
        
        if (!string.IsNullOrEmpty(oldComment) && string.IsNullOrEmpty(newComment))
            return TestConstants.RESULT_COMMENT_REMOVED;
        
        if (oldComment != newComment)
            return TestConstants.RESULT_COMMENT_MODIFIED;
        
        return TestConstants.RESULT_COMMENT_UPDATED;
    }

    private string CreateCommentHistoryText(string? oldComment, string? newComment)
    {
        // Normalize null/empty values for consistent comparison
        var normalizedOld = string.IsNullOrEmpty(oldComment) ? "" : oldComment;
        var normalizedNew = string.IsNullOrEmpty(newComment) ? "" : newComment;
        
        // If both are effectively empty, don't create history text
        if (normalizedOld == "" && normalizedNew == "")
            return "";
            
        if (string.IsNullOrEmpty(oldComment) && !string.IsNullOrEmpty(newComment))
            return newComment;
        
        if (!string.IsNullOrEmpty(oldComment) && string.IsNullOrEmpty(newComment))
            return $"Previous comment: {oldComment}";
        
        if (oldComment != newComment && !string.IsNullOrEmpty(oldComment))
            return $"New: {newComment} (Previous: {oldComment})";
        
        return newComment ?? "";
    }

    private IoUpdateDto CreateUpdateDto(Io io, string inspectorName)
    {
        return new IoUpdateDto
        {
            Id = io.Id,
            TestedBy = inspectorName,
            Result = io.Result,
            Comments = io.Comments,
            State = io.State,
            Version = io.Version, // Include current version for conflict detection
            Timestamp = io.Timestamp
        };
    }
} 

File: frontend/lib/services/io-test-service.ts
/**
 * IO Test Service
 *
 * Port of the C# IoTestService to TypeScript.
 * Handles recording test results (Pass/Fail/Reset), creating TestHistory audit records,
 * queuing results for cloud sync, and handling order mode (sequential testing).
 */

import { db } from '@/lib/db-sqlite'
import type { Io } from '@/lib/db-sqlite'
import { getPlcTags } from '../plc-client-manager'
import { checkInstallGate } from './install-gate'
import { getTagForIo, hasAnyMcm } from '../mcm-registry'

// Test result constants (matching C# TestConstants)
export const TEST_CONSTANTS = {
  RESULT_PASSED: 'Passed',
  RESULT_FAILED: 'Failed',
  RESULT_CLEARED: 'Cleared',
  RESULT_ADDRESSED: 'Addressed',
  RESULT_NOT_TESTED: 'Not Tested',
  RESULT_COMMENT_ADDED: 'Comment Added',
  RESULT_COMMENT_REMOVED: 'Comment Removed',
  RESULT_COMMENT_MODIFIED: 'Comment Modified',
  RESULT_COMMENT_UPDATED: 'Comment Updated',
  TIMESTAMP_FORMAT: 'MM/dd/yy h:mm:ss.fff a', // JavaScript date format
} as const

// Types
export interface IoWithState {
  id: number
  SubsystemId: number
  Name: string | null
  Description: string | null
  Result: string | null
  Timestamp: string | null
  Comments: string | null
  Order: number | null
  Version: number
  TagType?: string | null
  CloudSyncedAt?: string | null
  NetworkDeviceName?: string | null
  state?: string
}

export interface TestResultRequest {
  comments?: string
  currentUser?: string
  failureMode?: string
}

export interface CommentUpdateResult {
  success: boolean
  changesWereMade: boolean
  errorMessage?: string
}

export interface IoUpdateDto {
  id: number
  result?: string | null
  timestamp?: string | null
  comments?: string | null
  testedBy?: string
  state?: string
  version?: number
}

/**
 * Sanitize comment by removing HTML tags
 */
export function sanitizeComment(input: string | null | undefined): string | null {
  if (!input) return input ?? null
  return input.replace(/<[^>]*>/g, '')
}

/**
 * Create a timestamp in the standard format
 */
export function createTimestamp(): string {
  return new Date().toISOString()
}

/**
 * Get current PLC state for an IO by ID.
 *
 * Multi-MCM aware — if any MCM is registered (central-tool path), we resolve
 * the tag through mcm-registry so each IO addresses its owning controller.
 * If the registry is empty we fall back to the legacy single-PLC singleton.
 */
export function getPlcStateForIo(ioId: number): string | undefined {
  if (hasAnyMcm()) {
    const tag = getTagForIo(ioId)
    if (tag) return tag.state ?? undefined
    // intentional fall-through: registry knows nothing about this IO yet
    // (e.g. tags not loaded for the owning MCM), so try the singleton too
  }
  const { tags } = getPlcTags()
  const tag = tags.find(t => t.id === ioId)
  return tag?.state
}

/**
 * Get the next untested IO in order mode
 */
export function getNextUntestedIo(subsystemId: number): IoWithState | null {
  const io = db.prepare(
    'SELECT * FROM Ios WHERE SubsystemId = ? AND Result IS NULL ORDER BY "Order" ASC LIMIT 1'
  ).get(subsystemId) as Io | undefined

  if (!io) return null

  return {
    ...io,
    state: getPlcStateForIo(io.id)
  }
}

// Keep async signature for backward compat
export async function getNextUntestedIoAsync(subsystemId: number): Promise<IoWithState | null> {
  return getNextUntestedIo(subsystemId)
}

/**
 * Mark an IO test as passed
 */
export async function markTestPassedAsync(
  ioId: number,
  options: TestResultRequest = {}
): Promise<{ success: boolean; error?: string }> {
  return updateTestResult(ioId, TEST_CONSTANTS.RESULT_PASSED, options)
}

/**
 * Mark an IO test as failed
 */
export async function markTestFailedAsync(
  ioId: number,
  options: TestResultRequest = {}
): Promise<{ success: boolean; error?: string }> {
  return updateTestResult(ioId, TEST_CONSTANTS.RESULT_FAILED, options)
}

/**
 * Clear/reset an IO test result
 */
export async function clearTestResultAsync(
  ioId: number,
  currentUser: string = 'Unknown'
): Promise<{ success: boolean; error?: string }> {
  try {
    const io = db.prepare('SELECT * FROM Ios WHERE id = ?').get(ioId) as Io | undefined

    if (!io) {
      return { success: false, error: 'IO not found' }
    }

    // Check if already cleared
    const hadComments = !!io.Comments
    const hadResult = !!io.Result

    if (!hadComments && !hadResult) {
      // Already cleared, nothing to do
      return { success: true }
    }

    // Build history comment
    let historyComment: string | null = null
    if (hadResult && hadComments) {
      historyComment = io.Comments
    } else if (hadResult) {
      historyComment = `Cleared ${io.Result} result`
    } else {
      historyComment = 'Cleared comments'
    }

    const plcState = getPlcStateForIo(ioId)
    const timestamp = createTimestamp()

    // Use transaction for atomicity
    const clearTransaction = db.transaction(() => {
      db.prepare(
        'INSERT INTO TestHistories (IoId, Result, Timestamp, Comments, State, TestedBy) VALUES (?, ?, ?, ?, ?, ?)'
      ).run(ioId, TEST_CONSTANTS.RESULT_CLEARED, timestamp, historyComment, plcState ?? null, currentUser)

      db.prepare(
        'UPDATE Ios SET Result = NULL, Timestamp = NULL, Comments = NULL, Version = Version + 1 WHERE id = ?'
      ).run(ioId)
    })

    clearTransaction()

    return { success: true }
  } catch (error) {
    console.error('Error clearing test result:', error)
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error'
    }
  }
}

/**
 * Update comment for an IO
 */
export async function updateCommentAsync(
  ioId: number,
  newComment: string,
  currentUser: string = 'Unknown'
): Promise<CommentUpdateResult> {
  try {
    const io = db.prepare('SELECT * FROM Ios WHERE id = ?').get(ioId) as Io | undefined

    if (!io) {
      return { success: false, changesWereMade: false, errorMessage: 'IO not found' }
    }

    const oldComment = io.Comments ?? ''
    const normalizedNew = newComment ?? ''

    // No change needed
    if (oldComment === normalizedNew) {
      return { success: true, changesWereMade: false }
    }

    const plcState = getPlcStateForIo(ioId)
    const timestamp = createTimestamp()
    const { historyResult, historyComment } = determineCommentChange(oldComment, normalizedNew)

    // Use transaction for atomicity
    const commentTransaction = db.transaction(() => {
      db.prepare(
        'UPDATE Ios SET Comments = ?, Timestamp = ?, Version = Version + 1 WHERE id = ?'
      ).run(normalizedNew, timestamp, ioId)

      db.prepare(
        'INSERT INTO TestHistories (IoId, Result, Timestamp, Comments, State, TestedBy) VALUES (?, ?, ?, ?, ?, ?)'
      ).run(ioId, historyResult, timestamp, historyComment, plcState ?? null, currentUser)
    })

    commentTransaction()

    return { success: true, changesWereMade: true }
  } catch (error) {
    console.error('Error updating comment:', error)
    return {
      success: false,
      changesWereMade: false,
      errorMessage: error instanceof Error ? error.message : 'Unknown error'
    }
  }
}

/**
 * Get IO by ID with current PLC state
 */
export function getIoById(ioId: number): IoWithState | null {
  const io = db.prepare('SELECT * FROM Ios WHERE id = ?').get(ioId) as Io | undefined

  if (!io) return null

  return {
    ...io,
    state: getPlcStateForIo(io.id)
  }
}

// Keep async signature for backward compat
export async function getIoByIdAsync(ioId: number): Promise<IoWithState | null> {
  return getIoById(ioId)
}

/**
 * Get all IOs for a subsystem with current PLC states. Routes through
 * mcm-registry first when populated so each IO's state comes from the right
 * controller; falls back to the legacy singleton's tag list otherwise.
 */
export function getIosBySubsystem(subsystemId: number): IoWithState[] {
  const ios = db.prepare(
    'SELECT * FROM Ios WHERE SubsystemId = ? ORDER BY "Order" ASC'
  ).all(subsystemId) as Io[]

  return ios.map(io => ({
    ...io,
    state: getPlcStateForIo(io.id)
  }))
}

// Keep async signature for backward compat
export async function getIosBySubsystemAsync(subsystemId: number): Promise<IoWithState[]> {
  return getIosBySubsystem(subsystemId)
}

/**
 * Get test statistics for a subsystem
 */
export function getTestStats(subsystemId?: number): {
  total: number
  passed: number
  failed: number
  untested: number
} {
  if (subsystemId) {
    const total = (db.prepare('SELECT COUNT(*) as count FROM Ios WHERE SubsystemId = ?').get(subsystemId) as any).count
    const passed = (db.prepare('SELECT COUNT(*) as count FROM Ios WHERE SubsystemId = ? AND Result = ?').get(subsystemId, TEST_CONSTANTS.RESULT_PASSED) as any).count
    const failed = (db.prepare('SELECT COUNT(*) as count FROM Ios WHERE SubsystemId = ? AND Result = ?').get(subsystemId, TEST_CONSTANTS.RESULT_FAILED) as any).count
    return { total, passed, failed, untested: total - passed - failed }
  }

  const total = (db.prepare('SELECT COUNT(*) as count FROM Ios').get() as any).count
  const passed = (db.prepare('SELECT COUNT(*) as count FROM Ios WHERE Result = ?').get(TEST_CONSTANTS.RESULT_PASSED) as any).count
  const failed = (db.prepare('SELECT COUNT(*) as count FROM Ios WHERE Result = ?').get(TEST_CONSTANTS.RESULT_FAILED) as any).count
  return { total, passed, failed, untested: total - passed - failed }
}

// Keep async signature for backward compat
export async function getTestStatsAsync(subsystemId?: number): Promise<{
  total: number
  passed: number
  failed: number
  untested: number
}> {
  return getTestStats(subsystemId)
}

/**
 * Get test history for an IO
 */
export function getTestHistory(ioId: number, limit: number = 100) {
  return db.prepare(
    'SELECT * FROM TestHistories WHERE IoId = ? ORDER BY Timestamp DESC LIMIT ?'
  ).all(ioId, limit)
}

// Keep async signature for backward compat
export async function getTestHistoryAsync(ioId: number, limit: number = 100) {
  return getTestHistory(ioId, limit)
}

// Private helper functions

function updateTestResult(
  ioId: number,
  result: string,
  options: TestResultRequest = {}
): { success: boolean; error?: string } {
  try {
    const io = db.prepare('SELECT * FROM Ios WHERE id = ?').get(ioId) as Io | undefined

    if (!io) {
      return { success: false, error: 'IO not found' }
    }

    // Server-side authoritative gate. The UI also disables the buttons, but
    // a stale client / direct API hit would bypass that — this is the only
    // place that's safe.
    const gate = checkInstallGate(io)
    if (!gate.allowed) {
      return { success: false, error: gate.error }
    }

    const sanitizedComments = sanitizeComment(options.comments) ?? null
    const plcState = getPlcStateForIo(ioId)
    const timestamp = createTimestamp()

    // Use transaction for atomicity
    const testTransaction = db.transaction(() => {
      // A new Pass/Fail result is a re-test outcome — it supersedes any resolver
      // workflow state (Addressed / Clarification), so clear it. This is what
      // moves a re-checked item out of the Addressed bucket back to Pass/Fail.
      db.prepare(
        'UPDATE Ios SET Result = ?, Timestamp = ?, Comments = ?, PunchlistStatus = NULL, ClarificationNote = NULL, Version = Version + 1 WHERE id = ?'
      ).run(result, timestamp, sanitizedComments, ioId)

      db.prepare(
        'INSERT INTO TestHistories (IoId, Result, Timestamp, Comments, State, TestedBy, FailureMode) VALUES (?, ?, ?, ?, ?, ?, ?)'
      ).run(ioId, result, timestamp, io.Comments, plcState ?? null, options.currentUser ?? 'Unknown', options.failureMode || null)
    })

    testTransaction()

    return { success: true }
  } catch (error) {
    console.error('Error updating test result:', error)
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error'
    }
  }
}

function determineCommentChange(
  oldComment: string | null | undefined,
  newComment: string | null | undefined
): { historyResult: string; historyComment: string } {
  const normalizedOld = oldComment ?? ''
  const normalizedNew = newComment ?? ''

  const historyResult = determineCommentChangeType(normalizedOld, normalizedNew)
  const historyComment = createCommentHistoryText(normalizedOld, normalizedNew)

  return { historyResult, historyComment }
}

function determineCommentChangeType(oldComment: string, newComment: string): string {
  if (!oldComment && newComment) {
    return TEST_CONSTANTS.RESULT_COMMENT_ADDED
  }
  if (oldComment && !newComment) {
    return TEST_CONSTANTS.RESULT_COMMENT_REMOVED
  }
  if (oldComment !== newComment) {
    return TEST_CONSTANTS.RESULT_COMMENT_MODIFIED
  }
  return TEST_CONSTANTS.RESULT_COMMENT_UPDATED
}

function createCommentHistoryText(oldComment: string, newComment: string): string {
  // If both are empty, no history text needed
  if (!oldComment && !newComment) {
    return ''
  }

  if (!oldComment && newComment) {
    return newComment
  }

  if (oldComment && !newComment) {
    return `Previous comment: ${oldComment}`
  }

  if (oldComment !== newComment) {
    return `New: ${newComment} (Previous: ${oldComment})`
  }

  return newComment
}
